src/bespokelabs/curator/request_processor/base_online_request_processor.py

```diff
@@ -0,0 +1,467 @@
+
+    def has_capacity(self, token_estimate: int) -> bool:
+        """Check if there's enough capacity for a request"""
+        self.update_capacity()
+        has_capacity = (
+            self.available_request_capacity >= 1 and self.available_token_capacity >= token_estimate
+        )
+        if not has_capacity:
+            logger.debug(
+                f"No capacity for request with {token_estimate} tokens. "
+                f"Available capacity: {self.available_token_capacity} tokens, "
+                f"{self.available_request_capacity} requests."
+            )
+        return has_capacity
+
+    def consume_capacity(self, token_estimate: int):
+        """Consume capacity for a request"""
+        self.available_request_capacity -= 1
+        self.available_token_capacity -= token_estimate
+
+
+@dataclass
+class APIRequest:
+    """Stores an API request's inputs, outputs, and other metadata."""
+
+    task_id: int
+    generic_request: GenericRequest
+    api_specific_request: dict
+    attempts_left: int
+    result: list = field(default_factory=list)
+    prompt_formatter: PromptFormatter = field(default=None)
+    created_at: datetime.datetime = field(default_factory=datetime.datetime.now)
+
+
+class BaseOnlineRequestProcessor(BaseRequestProcessor, ABC):
+    """Abstract base class for online request processors that make real-time API calls."""
+
+    def __init__(
+        self,
+        model: str,
+        temperature: Optional[float] = None,
+        top_p: Optional[float] = None,
+        presence_penalty: Optional[float] = None,
+        frequency_penalty: Optional[float] = None,
+    ):
+        super().__init__(batch_size=None)
+        self.model: str = model
+        self.temperature: float | None = temperature
+        self.top_p: float | None = top_p
+        self.presence_penalty: float | None = presence_penalty
+        self.frequency_penalty: float | None = frequency_penalty
+        self.prompt_formatter: Optional[PromptFormatter] = None
+
+    @abstractmethod
+    def estimate_total_tokens(self, messages: list) -> int:
+        """Estimate total tokens for a request"""
+        pass
+
+    @abstractmethod
+    def estimate_output_tokens(self) -> int:
+        """Estimate output tokens for a request"""
+        pass
+
+    def check_structured_output_support(self) -> bool:
+        """Check if the model supports structured output"""
+        return True
+
+    def run(
+        self,
+        dataset: Optional[Dataset],
+        working_dir: str,
+        parse_func_hash: str,
+        prompt_formatter: PromptFormatter,
+    ) -> Dataset:
+        """Run completions using the online API with async processing."""
+        logger.info(f"Running {self.__class__.__name__} completions with model: {self.model}")
+
+        self.prompt_formatter = prompt_formatter
+        if self.prompt_formatter.response_format:
+            if not self.check_structured_output_support():
+                raise ValueError(
```

Comment:
Check litellm.AuthenticationError to give user feedback if their API key is not set or correct

Reply:
fixed. 

note here that it actually raises a 

instructor.exceptions.InstructorRetryException: litellm.AuthenticationError: Missing Anthropic API Key - A call is being made to anthropic but no key is set either in the environment variables or via params. Please set `ANTHROPIC_API_KEY` in your environment vars

because of the instructor wraps the exception, and within the exception, there's litellm.AuthenticationError in the error message, so I check for litellm.AuthenticationError in the exception message.